src/typed-ast/expr.h

```diff
@@ -0,0 +1,637 @@
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("float-literal").field(value).finish();
+    }
+  };
+
+  /// This represents a boolean literal, i.e. 'true' or 'false'.
+  class BooleanLiteral final : public Expr
+  {
+    bool value;
+
+  public:
+    explicit BooleanLiteral(SourceLocation location, bool value)
+    : Expr(Kind::BooleanLiteral, location), value(value)
+    {}
+
+    bool getValue()
+    {
+      return value;
+    }
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::BooleanLiteral;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("bool-literal").field(value).finish();
+    }
+  };
+
+  class StringLiteral final : public Expr
+  {
+    std::string value;
+
+  public:
+    explicit StringLiteral(SourceLocation location, std::string value)
+    : Expr(Kind::StringLiteral, location), value(value)
+    {}
+
+    const std::string& getValue() const
+    {
+      return value;
+    }
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::StringLiteral;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("string-literal").field(value).finish();
+    }
+  };
+
+  class InterpolateExpr final : public Expr
+  {
+    std::vector<ExprPtr> elements;
+
+  public:
+    explicit InterpolateExpr(
+      SourceLocation location, std::vector<ExprPtr> elements)
+    : Expr(Kind::Interpolate, location), elements(std::move(elements))
+    {}
+
+    std::vector<ExprPtr>& getElements()
+    {
+      return elements;
+    }
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::Interpolate;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("seq").field(elements).finish();
+    }
+  };
+
+  class TupleExpr final : public Expr
+  {
+    std::vector<ExprPtr> elements;
+
+  public:
+    explicit TupleExpr(SourceLocation location, std::vector<ExprPtr> elements)
+    : Expr(Kind::Tuple, location), elements(std::move(elements))
+    {}
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::Tuple;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("tuple").field(elements).finish();
+    }
+  };
+
+  class WhenExpr final : public Expr
+  {
+    std::vector<ExprPtr> arguments;
+    ExprPtr body;
+
+  public:
+    explicit WhenExpr(
+      SourceLocation location, std::vector<ExprPtr> arguments, ExprPtr body)
+    : Expr(Kind::When, location),
+      arguments(std::move(arguments)),
+      body(std::move(body))
+    {}
+
+    static bool classof(const Expr* E)
+    {
+      return E->getKind() == Kind::When;
+    }
+
+    void print(NodePrinter& out) const override
+    {
+      out.begin("when").field(arguments).field(body).finish();
+    }
+  };
+
+  class NewExpr final : public Expr
+  {
+    std::vector<std::unique_ptr<MemberDef>> elements;
```

Comment:
Do you really mean `MemberDef` or `FieldDef`?

Reply:
`MemberDef`. You are allowed to do, for example:

```ts
interface I {
  foo();
}

new I {
  foo() { ... }
}
```

Or at least that's how I understand the grammar.

Reply:
Hm, this looks weird to me, but so did a lot of Verona code before. Should be fine for now.